### PR TITLE
Refactor BucketBatch to BucketBatcher

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63185 Refactor BucketBatch**
* #63184 Replace str by repr for DataChunk

- To land BucketBatch in `torchdata`, functional API needs to be removed.
- Refactor the BucketBatch to be same as `torchdata`
- Add deprecation warning to users
- Add corresponding tests

Differential Revision: [D30288893](https://our.internmc.facebook.com/intern/diff/D30288893)

### CODE REVIEW

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30288893).

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30288893).

File: test/test_datapipe.py [hunk 1]
```diff
@@ -711,52 +711,48 @@ def test_unbatch_datapipe(self):
...
+                    bucket = []
+            _verify_bucket_sorted(bucket)
+
         def _sort_fn(data):
-            return data
+            return sorted(data)
+
+        # In-batch shuffle
+        _helper(batch_size=3, drop_last=False, batch_num=5, sort_key=_sort_fn)
+        #  _helper(batch_size=3, drop_last=False, batch_num=2, bucket_num=2, sort_key=_sort_fn)
```
Comment (VitalyFedyunin) [inline test/test_datapipe.py:752]: nit: Drop comments?

File: torch/utils/data/datapipes/iter/grouping.py [hunk 2]
```diff
@@ -132,9 +133,13 @@ def _dive(self, element, unbatch_level):
             else:
                 raise IndexError(f"unbatch_level {self.unbatch_level} exceeds the depth of the DataPipe")
 
+# TODO(Erjia): https://github.com/pytorch/pytorch/issues/63095
```
Comment (VitalyFedyunin) [inline torch/utils/data/datapipes/iter/grouping.py:136]: super nit: better TODO(accout)

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by ejguan** [comment]
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30288893).

**Pull request closed without merging**